File: nodeJs/tic-tac-toe-board-game/game.js
'use strict';
let Player = require('./player');
let Board = require('./board');
let ResultAnalyze = require('./resultAnalyze');
let Mark = require('./mark');
let Result = require('./result');

class Game {

    constructor(player1, player2) {
        if (player1 == null || player2 == null)
            throw new NullPointerException();
        this._player1 = player1;
        this._player2 = player2;
        this._board = null;
        this._resultAnalyze = null;
        this._playerSign = null;
        this._result = null;
    }
    startGame() {
        this._board = new Board();
        this._resultAnalyze = new ResultAnalyze(this._board);
        this._playerSign = this._player1.mark;
    }

    setMarkInCellOfBoard(index) {

        this._board.setBoardCellMark(index, this._playerSign);
        this._playerSign = this.changePlayerSign();

    }

    changePlayerSign() {
        if (this._playerSign == Mark.X)
            return Mark.O;
        if (this._playerSign == Mark.O)
            return Mark.X;
        return null;
    }

    getCurrentPlayer() {
        if (this._player1.mark == this._playerSign)
            return this._player1;
        return this._player2;
    }
    displayBoard() {
        this._board.displayBoard();
    }

    getResult() {

        this._result = this._resultAnalyze.genrateResultState();
        if (this._result == Result.WIN)
            return this.checkWhichPlayerIsWin();

        return this._result.toString();
    }

    checkWhichPlayerIsWin() {
        if (this._player1.mark.toString() == this._resultAnalyze.getWiningPlayerSign().toString())
            return this._player1.name + " is Win";
        return this._player2.name + " is Win";
    }
}
module.exports = Game;

File: OOP/tic-tac-toe-board-game-app/src/com/techlab/game/Game.java
package com.techlab.game;

import com.techlab.Exception.AllReadyMarkCellException;
import com.techlab.board.Board;
import com.techlab.board.Mark;
import com.techlab.player.Player;
import com.techlab.result.analyze.Result;
import com.techlab.result.analyze.ResultAnalyze;

public class Game {
	private Player player1;
	private Player player2;
	private Board board;
	private ResultAnalyze resultAnalyze;
	private Mark playerSign;

	public Game(Player player1, Player player2) {
		if (player1.equals(null) || player2.equals(null)) {
			throw new NullPointerException();
		}
		this.player1 = player1;
		this.player2 = player2;
	}

	public void startGame() {
		board = new Board();
		resultAnalyze = new ResultAnalyze(board);
		playerSign = player1.getMark();
	}

	public void setMarkInCellOfBoard(int position) throws AllReadyMarkCellException {

		board.setBoardCellMark(position, playerSign);
		playerSign = changePlayerSign();

	}

	private Mark changePlayerSign() {
		if (playerSign.equals(Mark.X))
			return Mark.O;
		if (playerSign.equals(Mark.O))
			return Mark.X;
		return null;
	}

	public Player getCurrentPlayer() {
		if (player1.getMark().equals(playerSign))
			return player1;
		return player2;
	}

	public String getResult() {

		Result result = resultAnalyze.genrateResultState();
		if (result.equals(Result.WIN)) {
			return checkWhichPlayerIsWin();
		}
		return result.toString();
	}

	private String checkWhichPlayerIsWin() {
		if (player1.getMark().toString().equals(resultAnalyze.getWiningPlayerSign().toString())) {
			return player1.getName() + " is Win";
		}
		return player2.getName() + " is Win";
	}

}
